pkg/agent/http_api_context.go

```diff
@@ -0,0 +1,248 @@
+	sync.Once
+
+	tag               string
+	response          http.ResponseWriter
+	request           *http.Request
+	isMe              bool
+	proxy             func() error
+	ignoreReaderLabel func(string) bool
+	ownedNamespaces   func() (hasAll bool, namespaceSet data.Set)
+	remoteQuery       func(query string) (result prommodel.Value, err error)
+}
+
+type jsonResponseData struct {
+	Status    string      `json:"status"`
+	Data      interface{} `json:"data,omitempty"`
+	ErrorType string      `json:"errorType,omitempty"`
+	Error     string      `json:"error,omitempty"`
+}
+
+func (c *apiContext) responseJSON(data interface{}) (err error) {
+	c.Do(func() {
+		resp := c.response
+		resp.Header().Set(httputil.ContentTypeHeader, httputil.JSONContentType)
+
+		responseData := &jsonResponseData{
+			Status: "success",
+			Data:   data,
+		}
+
+		json := jsoniter.ConfigCompatibleWithStandardLibrary
+		respBytes, marshalErr := json.Marshal(responseData)
+		if marshalErr != nil {
+			err = errors.Wrap(marshalErr, internalErr)
+			return
+		}
+
+		if _, writeErr := resp.Write(respBytes); writeErr != nil {
+			err = errors.Wrap(writeErr, internalErr)
+		}
+	})
+
+	return
+}
+
+func (c *apiContext) responseProto(data proto.Message) (err error) {
+	c.Do(func() {
+		resp := c.response
+		resp.Header().Set(httputil.ContentTypeHeader, httputil.ProtoContentType)
+		resp.Header().Set(httputil.ContentEncodingHeader, "snappy")
+
+		if data == nil {
+			resp.WriteHeader(http.StatusNoContent)
+			return
+		}
+
+		responseData, marshalErr := proto.Marshal(data)
+		if marshalErr != nil {
+			err = errors.Wrap(marshalErr, internalErr)
+			return
+		}
+
+		respBytes := snappy.Encode(nil, responseData)
+		if _, writeErr := resp.Write(respBytes); writeErr != nil {
+			err = errors.Wrap(writeErr, internalErr)
+		}
+	})
+
+	return
+}
+
+func (c *apiContext) responseMetrics(data *promgo.MetricFamily) (err error) {
+	c.Do(func() {
+		req, resp := c.request, c.response
+
+		respFormat := expfmt.Negotiate(req.Header)
+		respEncoder := expfmt.NewEncoder(resp, respFormat)
+		resp.Header().Set(httputil.ContentTypeHeader, string(respFormat))
+
+		if data == nil {
+			return
+		}
+
+		if encodeErr := respEncoder.Encode(data); encodeErr != nil {
+			err = errors.Wrap(encodeErr, internalErr)
+		}
+	})
+
+	return
+}
+
+type apiContextPool struct {
+	pool *sync.Pool
+}
+
+func (p *apiContextPool) grant(accessToken string, config *agentConfig, ownedNamespacesViewer kubeauth.OwnedNamespacesViewer, remoteAPI promapiv1.API, proxy http.Handler, resp http.ResponseWriter, req *http.Request) (*http.Request, *apiContext) {
+	apiCtx := p.pool.Get().(*apiContext)
+
+	newReqCtx := context.WithValue(req.Context(), apiContextKey, apiCtx)
+	newReq := req.WithContext(newReqCtx)
+
+	apiCtx.tag = fmt.Sprintf("%016x", time.Now().Unix())
+	apiCtx.request = newReq
+	apiCtx.response = resp
+	apiCtx.isMe = config.myToken == accessToken
+	apiCtx.proxy = func() (err error) {
+		apiCtx.Do(func() {
```

Comment:
Why using `once.Do` for `proxy.ServeHTTP`? Since you use pool for apiContext object, `apiCtx.proxy()` will do nothing if the apiContext object is reused.

Reply:
Use `sync.Once` to ensure only call once in `apiCtx.proxy()`, `apiCtx.responseXXX()`

Reply:
But when it reuse the `apiContext` Object in the pool, `apiCtx.proxy()` would not be called again and will not get any response for the new request.